# Spazi Vettoriali e Parole

## Che cos'è un Vettore

Immaginiamo di trovarci su un'isola immaginaria, come mostrano le immagini allegate. Su quest’isola ci sono varie attrazioni turistiche: un faro, un vulcano, una palma, una spiaggia, un castello. Sono luoghi ben precisi, visibili e facilmente riconoscibili. Ora proviamo a fare un piccolo sforzo mentale: vogliamo **descrivere la posizione** di ciascuna attrazione in un modo chiaro, condivisibile e, soprattutto, **calcolabile**.

<img src="./pictures/vettori-1.png" alt="Descrizione dell'immagine" width="400" height="200" />

### Dalle immagini alle coordinate

Nel primo disegno, notiamo che l'isola è attraversata da due linee perpendicolari: una orizzontale (asse X, Est-Ovest) e una verticale (asse Y, Nord-Sud). Questo sistema di assi divide l’isola in quattro quadranti, e permette di assegnare a ogni luogo una **coppia di numeri**: le sue **coordinate**.

Ad esempio:

* Il **Faro** si trova a sinistra e un po’ in alto, in posizione (−3, 2)
* La **Spiaggia** è sulla destra e un po’ in basso, in (4, −1)
* Il **Castello** è a destra e in basso, in (3, −2)


<img src="./pictures/vettori-2.png" alt="Descrizione dell'immagine" width="400" height="200" />

### Dalle coordinate al concetto di vettore

Finora abbiamo parlato di “punti” su una mappa. Ma se ora immaginiamo di voler **muoverci** da un punto all’altro — ad esempio dal faro al vulcano — abbiamo bisogno di qualcosa che non dica solo **dove siamo**, ma anche **come spostarci**.

Qui entra in gioco il **vettore**.

Un **vettore** è una **freccia che unisce due punti**, indicando:

1. **direzione** (verso dove andare)
2. **verso** (da dove a dove)
3. **lunghezza** (quanto spostarsi)

Ad esempio, il vettore che va dal Faro al Vulcano dice: “spostati di 7 unità verso Est (da −4 a 3) e di 1 unità verso Nord (da 2 a 3)”. Questo lo possiamo scrivere come il vettore **(7, 1)**.



**Il salto concettuale: spazio vettoriale**

Quando iniziamo a lavorare con vettori, ci accorgiamo che hanno delle proprietà molto interessanti:

* Possiamo **sommarli** (fare un tragitto in due tappe)
* Possiamo **scalarli** (fare lo stesso tragitto ma due volte più lungo)
* Possiamo **confrontare** vettori che hanno la stessa direzione, verso e lunghezza, anche se partono da punti diversi

Tutto questo avviene in quello che in matematica si chiama **spazio vettoriale**: un insieme di vettori su cui è possibile fare queste operazioni in modo coerente.

Nel nostro caso, lo spazio è il piano delle coordinate (X, Y), dove ogni vettore è rappresentato da una coppia di numeri. È come un linguaggio universale per descrivere spostamenti e posizioni.


**In sintesi**

* Ogni attrazione turistica dell’isola ha una **posizione**, data da una **coppia di coordinate** (x, y)
* Ogni **spostamento** da un punto a un altro può essere rappresentato come un **vettore**
* I vettori hanno direzione, verso e lunghezza
* Insieme, i vettori formano uno **spazio vettoriale**, una struttura matematica che ci permette di analizzare e combinare movimenti e relazioni spaziali

### Un primo passo verso l'astrazione: misurare caratteristiche non spaziali

Abbiamo compreso facilmente che un vettore nello spazio tridimensionale è rappresentato da tre numeri (le coordinate cartesiane x, y, z). Ad esempio:

* La posizione di una città può essere identificata da tre coordinate geografiche.
* Un punto in una stanza può essere identificato da tre coordinate (lunghezza, larghezza, altezza).

In questo caso le coordinate rappresentano letteralmente una posizione nello spazio fisico, quello a cui siamo abituati.

Adesso facciamo un piccolo passo verso l’astrazione, usando sempre dei vettori con coordinate numeriche, ma stavolta non riferite a posizioni nello spazio. Per esempio, immaginiamo di voler rappresentare delle persone attraverso alcune caratteristiche misurabili numericamente:

> **Persona → (età, peso, altezza)**

In questo modo, una persona può essere vista come un "punto" in uno spazio astratto definito da queste tre dimensioni: età, peso e altezza. Ad esempio:

* Mario: (30 anni, 75 kg, 180 cm)
* Lucia: (28 anni, 62 kg, 165 cm)

***In questo "spazio delle persone", due individui con caratteristiche simili (ad esempio età simile, altezza simile) saranno due "punti" vicini tra loro***.

Notiamo che, anche se parliamo ancora di numeri semplici, qui abbiamo fatto un primo passo verso l’astrazione: non siamo più nello spazio fisico, ma in uno spazio di **caratteristiche numeriche**.


### Un ulteriore passo verso l’astrazione: includere caratteristiche non numeriche (qualitative)

Ora facciamo un altro piccolo passo avanti: aggiungiamo caratteristiche non immediatamente numeriche ma che possiamo rappresentare numericamente.

Immaginiamo ad esempio di voler rappresentare dei film tramite vettori numerici, usando alcune caratteristiche come:

* Quanto è comico (da 0 a 10)
* Quanto è drammatico (da 0 a 10)
* Quanto è romantico (da 0 a 10)

Ogni film potrebbe essere descritto da un vettore di tre numeri che rappresentano intensità di caratteristiche qualitative:

* Film A (molto comico, poco drammatico, medio romantico): `(8, 2, 5)`
* Film B (poco comico, molto drammatico, molto romantico): `(1, 9, 8)`

In questo spazio astratto, due film simili si troveranno "vicini" tra loro, proprio come due città vicine in una mappa, anche se lo spazio non ha più nulla a che vedere con posizioni fisiche.

### Il salto finale: rappresentare concetti astratti come le parole

Finalmente, siamo pronti per il salto finale, quello più astratto di tutti:

> E se provassimo a rappresentare **il significato delle parole** con numeri?

L'idea del word embedding nasce proprio qui: rappresentare una parola come una lista di numeri (un vettore), ciascuno dei quali esprime quanto la parola è associata a concetti o contesti particolari.

Ad esempio, per le parole "gatto", "cane" e "automobile", un modello intelligente potrebbe assegnare coordinate numeriche in modo che:

* `"gatto" = (8, 9, 1, ...)`
* `"cane" = (7.5, 9.2, 1.5, ...)`
* `"automobile" = (0.5, 1, 9.5, ...)`

Anche se queste coordinate non sono più interpretabili facilmente una ad una (sono prodotte automaticamente dai modelli di AI), si mantengono due proprietà importanti:

* Parole con significati simili hanno vettori vicini nello spazio astratto.
* Parole con significati molto diversi hanno vettori lontani.

Quindi, come nello spazio ordinario:

* Vicinanza geografica → città vicine.
* Vicinanza astratta → significati vicini.

### Riassunto del percorso fatto:

| Livello                | Esempio                 | Componenti del vettore                                                |
| :---------------------- | :----------------------- | :--------------------------------------------------------------------- |
| 1️⃣ Ordinario          | Posizione geografica    | Coordinate spaziali (x,y,z)                                           |
| 2️⃣ Prima astrazione   | Persone                 | Caratteristiche numeriche (età, peso, altezza)                        |
| 3️⃣ Seconda astrazione | Film                    | Caratteristiche qualitative numeriche (comico, romantico, drammatico) |
| 4️⃣ Totale astrazione  | Parole (Word embedding) | Coordinate semantiche automatiche                                     |

## Il modello Bag of Words

Immaginiamo che la nostra macchina capisca solo un piccolo dizionario di poche parole. Per costruire i vettori che rappresentano le frasi date utilizzando il metodo *Bag of Words*, inizieremo creando un dizionario con questi termini. Nella *Bag of Words* ogni frase è trasformata in un vettore, lungo esattamente come il vocabolario, nel quale ad ogni elemento corrisponde una parola del dizionario.

Il valore di ciascun elemento del vettore verrà calcolato contando la frequenza di ciascuna parola. Se una parola del dizionario non appare nella frase il suo conteggio sarà $0$. Se appare una o più volte, il valore sarà pari a $1$ o al numero totale di volte che la parola appare nella frase...

In [1]:
def costruisci_dizionario(frase):
    import re
    from collections import Counter

    # Pulisce e divide in parole
    parole = re.findall(r'\b\w+\b', frase.lower())
    conteggio = Counter(parole)

    # Lista predefinita di parole da visualizzare
    parole_uniche = ['amo', 'casa', 'il', 'ma', 'non', 'sole', 'stare', 'viaggiare', 'volare', 'vorrei']

    # Larghezza fissa per ciascuna colonna
    larghezza = max(len(w) for w in parole_uniche) + 2
    n = len(parole_uniche)

    # Funzioni per costruire la tabella
    def linea_superiore():
        return "┌" + "┬".join(["─" * larghezza] * n) + "┐"

    def linea_centrale():
        return "├" + "┼".join(["─" * larghezza] * n) + "┤"

    def linea_inferiore():
        return "└" + "┴".join(["─" * larghezza] * n) + "┘"

    def riga_valori(valori):
        return "│" + "│".join(f"{valori[i]:^{larghezza}}" for i in range(n)) + "│"

    # Intestazione
    dizionario_str = "VETTORE\n"
    dizionario_str += "Indice:\n"
    dizionario_str += " " + " ".join(f"{i+1:^{larghezza}}" for i in range(n)) + "\n\n"

    # Costruzione riga valori (parole)
    dizionario_str += linea_superiore() + "\n"
    dizionario_str += riga_valori(parole_uniche) + "\n"
    dizionario_str += linea_centrale() + "\n"

    # Riga dei conteggi o spazi vuoti
    if not parole:
        frequenze = [" " for _ in parole_uniche]
    else:
        frequenze = [conteggio[w] for w in parole_uniche]

    dizionario_str += riga_valori(frequenze) + "\n"
    dizionario_str += linea_inferiore()

    return dizionario_str

In [2]:
# ESEMPIO DI UTILIZZO
frase = ""
print('\nFRASE: ', frase, '\n')
print(costruisci_dizionario(frase))

frase = "amo viaggiare ma non amo volare"
print('\nFRASE: ', frase, '\n')
print(costruisci_dizionario(frase))

frase = "vorrei stare a casa"
print('\nFRASE: ', frase, '\n')
print(costruisci_dizionario(frase))


FRASE:   

VETTORE
Indice:
      1           2           3           4           5           6           7           8           9          10     

┌───────────┬───────────┬───────────┬───────────┬───────────┬───────────┬───────────┬───────────┬───────────┬───────────┐
│    amo    │   casa    │    il     │    ma     │    non    │   sole    │   stare   │ viaggiare │  volare   │  vorrei   │
├───────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┤
│           │           │           │           │           │           │           │           │           │           │
└───────────┴───────────┴───────────┴───────────┴───────────┴───────────┴───────────┴───────────┴───────────┴───────────┘

FRASE:  amo viaggiare ma non amo volare 

VETTORE
Indice:
      1           2           3           4           5           6           7           8           9          10     

┌───────────┬───────────┬───────────┬───────────┬──────────

Sembra funzionare, ogni frase ha il suo vettore. Ma quando proviamo frasi più complesse come le seguenti

In [3]:
frase = "amo viaggiare, vorrei non stare a casa"
print('\nFRASE: ', frase, '\n')
print(costruisci_dizionario(frase))

frase = "non amo viaggiare, vorrei stare a casa"
print('\nFRASE: ', frase, '\n')
print(costruisci_dizionario(frase))


FRASE:  amo viaggiare, vorrei non stare a casa 

VETTORE
Indice:
      1           2           3           4           5           6           7           8           9          10     

┌───────────┬───────────┬───────────┬───────────┬───────────┬───────────┬───────────┬───────────┬───────────┬───────────┐
│    amo    │   casa    │    il     │    ma     │    non    │   sole    │   stare   │ viaggiare │  volare   │  vorrei   │
├───────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────┤
│     1     │     1     │     0     │     0     │     1     │     0     │     1     │     1     │     0     │     1     │
└───────────┴───────────┴───────────┴───────────┴───────────┴───────────┴───────────┴───────────┴───────────┴───────────┘

FRASE:  non amo viaggiare, vorrei stare a casa 

VETTORE
Indice:
      1           2           3           4           5           6           7           8           9          10     

┌───────────┬─

ci accorgiamo che entrambe le frasi hanno la stessa rappresentazione vettoriale. Questo significa che, secondo questo modello, le frasi avranno il medesimo significato! Tuttavia le due frasi hanno significati opposti e questo dimostra chiaramente uno dei principali limiti di questo metodo nell'analisi del linguaggio naturale.

---

## Apprendere il Significato dal Contesto

Coscienti di questi limiti i ricercatori hanno sviluppato modelli che partono dall'assunto che le parole che appaiono frequentemente vicine le une alle altre hanno significati più strettamente collegati. Quindi l'ipotesi è che analizzando le parole all'interno di una certa finestra di contesto attorno ad una parola target il modello possa apprendere rappresentazioni più precise. 

---

### Cos'è un word embedding?

Un **word embedding** è un modo per rappresentare una parola come una **sequenza di numeri** (un vettore), così che i computer possano lavorare con le parole in modo simile a come fanno con i numeri.

Ma non sono numeri qualsiasi. A differenza del modello BoW, ogni numero nel vettore **"porta con sé un significato"**, perché è stato appreso analizzando milioni di frasi e testi. In altre parole, i vettori **catturano il significato** delle parole basandosi su come queste vengono usate nel linguaggio.

**Cosa vuol dire “dimensione” in un word embedding?**

Immagina che ogni parola venga trasformata in un vettore con, ad esempio, **300 numeri**. Ogni numero è una **dimensione** dell'embedding.

Ora: queste **dimensioni non sono etichette esplicite**, ma rappresentano **sfumature di significato** che l'algoritmo ha imparato da solo. Alcune di queste dimensioni possono (in modo implicito) rappresentare:

* il concetto di **maschile vs femminile**
* il grado di **astrazione** di una parola
* la sua **carica emotiva** (positiva/negativa)
* il legame con **luoghi** o **tempi**
* la **categoria grammaticale** (sostantivo, verbo...)

**Cosa si intende per "dimensione semantica"?**

Una **dimensione semantica** è quindi **una direzione nello spazio dei significati**, lungo la quale possiamo cogliere un cambiamento semantico specifico.

Un esempio molto noto:

* Se prendiamo i vettori delle parole `re` (king), `regina` (queen), `uomo` (man) e `donna` (woman), il famoso esempio di word embedding è:

  ```
  re − uomo + donna ≈ regina
  ```

In questo caso, la **differenza tra "re" e "uomo"** può essere interpretata come una **dimensione semantica di regalità**, e la **differenza tra "uomo" e "donna"** come una **dimensione semantica di genere**.

**Un'analogia visiva**

Immagina uno spazio tridimensionale:

* L’asse X potrebbe rappresentare **il grado di positività** della parola
* L’asse Y il **campo semantico** (es. "cibo", "emozione", "luogo")
* L’asse Z il **livello di concretezza** (oggetti tangibili vs concetti astratti)

Una parola come **"cioccolato"** potrebbe avere coordinate (8, 2, 9), mentre **"libertà"** potrebbe stare a (5, 8, 1).

In un vero word embedding ci sono **molte più dimensioni** (50, 100, 300...), ma il concetto è lo stesso: ogni direzione rappresenta **una possibile variazione di significato**.

**Riassumendo**

> Una **dimensione semantica** in un word embedding è una direzione nello spazio matematico che riflette **una caratteristica latente del significato** delle parole, appresa dai dati.

Non sappiamo sempre **cosa** rappresenta esattamente ogni dimensione, ma possiamo studiarle osservando come le parole si posizionano e si muovono nello spazio.

---

Il seguente script crea una **rappresentazione tridimensionale (3D)** di un piccolo **word embedding**, cioè una mappa spaziale in cui ogni parola è rappresentata da un punto nello spazio, in base a **tre “dimensioni semantiche”:**

1. **Vivente** (asse X)
2. **Dimensione fisica** (asse Y)
3. **Tecnologia** (asse Z)

**Cosa fa il codice, passo per passo**

1. **Definizione dei dati**

```python
parole = {
    "gatto"      : (1, 0.25, 0),
    "cane"       : (1, 0.5, 0),
    "elefante"   : (1, 1, 0),
    "drone"      : (0, 0.5, 1),
    "automobile" : (0, 1, 1),
    "camion"     : (0, 1.2, 1),
    "pietra"     : (0, 0, 0),
}
```

Qui si definisce un **mini dizionario di parole** in cui ad ogni parola viene associata una **tripla di numeri**, come se fosse un punto in uno spazio 3D.

* `gatto`, `cane`, `elefante` hanno X = 1 → sono **esseri viventi**
* `drone`, `automobile`, `camion` hanno Z = 1 → sono **tecnologici**
* `elefante` ha Y = 1 → è **grande**, mentre `pietra` ha tutto a 0 → non è vivente, non tecnologica, piccola

> 👉 Questo è un esempio didattico: in un vero word embedding queste coordinate verrebbero apprese da un algoritmo. Qui invece sono scelte a mano per illustrare il concetto.

2. **Separazione delle coordinate**

```python
x = [v[0] for v in parole.values()]
y = [v[1] for v in parole.values()]
z = [v[2] for v in parole.values()]
labels = list(parole.keys())
```

Qui si preparano tre liste (x, y, z) che raccolgono le **coordinate di tutte le parole**, separate per asse.

3. **Disegno degli assi**

```python
assi = [
    go.Scatter3d(...),  # asse X (Vivente)
    go.Scatter3d(...),  # asse Y (Dimensione)
    go.Scatter3d(...)   # asse Z (Tecnologia)
]
```

Si disegnano tre **assi cartesiani** in nero nello spazio 3D, così da mostrare il contesto spaziale delle parole.

4. **Disegno dei punti (parole)**

```python
punti = go.Scatter3d(
    x=x, y=y, z=z,
    mode='markers+text',
    text=labels,
    ...
)
```

Ogni parola viene visualizzata come:

* un **punto azzurro**
* con una **etichetta testuale** (il nome della parola)
* posizionata secondo le coordinate definite sopra

5. **Creazione e visualizzazione della scena**

```python
fig = go.Figure(data=assi + [punti])
fig.update_layout(...)
fig.show()
```

Infine si:

* combinano gli assi + i punti in un’unica figura
* si etichettano gli assi (`Vivente`, `Dimensione`, `Tecnologia`)
* si mostra la scena interattiva in 3D


**Cosa rappresenta questa visualizzazione?**

* Le **parole sono punti** in uno spazio dove la distanza e la posizione **riflettono caratteristiche semantiche**.
* Se due parole sono vicine, significa che **hanno significati simili** lungo le dimensioni scelte.
* Se sono lontane, rappresentano **concetti diversi**.

---
**Esempi interpretativi**

* `gatto` e `cane` sono vicini → entrambi viventi, taglia simile
* `camion` e `automobile` sono vicini → non viventi, grandi, tecnologici
* `pietra` è isolata → non è vivente, non è tecnologica, è piccola
* `drone` è vicino alle auto → tecnologico, ma più piccolo

In [11]:
import plotly.graph_objects as go

# Definizione delle parole e delle coordinate semantiche
parole = {
    "gatto": (1, 0.25, 0),
    "cane": (1, 0.5, 0),
    "elefante": (1, 1, 0),
    "drone": (0, 0.5, 1),
    "automobile": (0, 1, 1),
    "camion": (0, 1.2, 1),
    "pietra": (0, 0.1, 0),
}

# Coordinate separate
x = [v[0] for v in parole.values()]
y = [v[1] for v in parole.values()]
z = [v[2] for v in parole.values()]
labels = list(parole.keys())

# Tracciamento assi manuali in nero
assi = [
    go.Scatter3d(x=[0, 1.2], y=[0, 0], z=[0, 0], mode='lines', line=dict(color='black', width=4), showlegend=False),
    go.Scatter3d(x=[0, 0], y=[0, 1.2], z=[0, 0], mode='lines', line=dict(color='black', width=4), showlegend=False),
    go.Scatter3d(x=[0, 0], y=[0, 0], z=[0, 1.2], mode='lines', line=dict(color='black', width=4), showlegend=False)
]

# Punti e testi
punti = go.Scatter3d(
    x=x, y=y, z=z,
    mode='markers+text',
    text=labels,
    textposition='top center',
    marker=dict(size=3, color='skyblue', line=dict(width=1, color='darkblue'))
)

# Composizione della figura
fig = go.Figure(data=assi + [punti])

# Specifica le dimensioni della finestra di output
fig.update_layout(
    title='Word Embedding 3D (Vivente, Grandezza, Tecnologia)',
    width=600,   # larghezza in pixel
    height=500,  # altezza in pixel
    scene=dict(
        xaxis_title='Vivente',
        yaxis_title='Grandezza',
        zaxis_title='Tecnologia',
        xaxis=dict(showspikes=False),
        yaxis=dict(showspikes=False),
        zaxis=dict(showspikes=False)
    ),
    margin=dict(l=0, r=0, b=0, t=40)
)

fig.show()


In [8]:
import plotly.graph_objects as go
import numpy as np

# Definizione delle parole e delle coordinate semantiche con colori semantici
parole_data = {
    "gatto": {"coords": (1, 0.25, 0), "color": "#FF6B6B", "category": "Animale"},
    "cane": {"coords": (1, 0.5, 0), "color": "#FF8E53", "category": "Animale"},
    "elefante": {"coords": (1, 1, 0), "color": "#FF9F43", "category": "Animale"},
    "drone": {"coords": (0, 0.5, 1), "color": "#5F27CD", "category": "Tecnologia"},
    "automobile": {"coords": (0, 1, 1), "color": "#3742FA", "category": "Tecnologia"},
    "camion": {"coords": (0, 1.2, 1), "color": "#2F3542", "category": "Tecnologia"},
    "pietra": {"coords": (0, 0, 0), "color": "#A4B0BE", "category": "Oggetto"},
}

# Estrazione coordinate e metadati
x = [data["coords"][0] for data in parole_data.values()]
y = [data["coords"][1] for data in parole_data.values()]
z = [data["coords"][2] for data in parole_data.values()]
labels = list(parole_data.keys())
colors = [data["color"] for data in parole_data.values()]
categories = [data["category"] for data in parole_data.values()]

# Creazione assi con stile migliorato
def create_axis_line(start, end, name, color='#2C3E50'):
    """Crea una linea per gli assi con freccia"""
    return go.Scatter3d(
        x=[start[0], end[0]], 
        y=[start[1], end[1]], 
        z=[start[2], end[2]],
        mode='lines',
        line=dict(color=color, width=6),
        name=name,
        showlegend=False,
        hoverinfo='skip'
    )

# Creazione frecce per gli assi
assi = [
    create_axis_line([0, 0, 0], [1.3, 0, 0], 'Asse X - Vivente', '#E74C3C'),
    create_axis_line([0, 0, 0], [0, 1.3, 0], 'Asse Y - Dimensione', '#2ECC71'),
    create_axis_line([0, 0, 0], [0, 0, 1.3], 'Asse Z - Tecnologia', '#3498DB')
]

# Aggiunta etichette degli assi con posizionamento migliorato
axis_labels = [
    go.Scatter3d(x=[1.4], y=[0], z=[0], mode='text', text=['Vivente'], 
                textfont=dict(size=16, color='#E74C3C', family="Arial Black"), 
                showlegend=False, hoverinfo='skip'),
    go.Scatter3d(x=[0], y=[1.4], z=[0], mode='text', text=['Dimensione'], 
                textfont=dict(size=16, color='#2ECC71', family="Arial Black"), 
                showlegend=False, hoverinfo='skip'),
    go.Scatter3d(x=[0], y=[0], z=[1.4], mode='text', text=['Tecnologia'], 
                textfont=dict(size=16, color='#3498DB', family="Arial Black"), 
                showlegend=False, hoverinfo='skip')
]

# Punti principali con stile migliorato
punti_principali = go.Scatter3d(
    x=x, y=y, z=z,
    mode='markers+text',
    text=labels,
    textposition='top center',
    textfont=dict(
        size=14,
        color='#2C3E50',
        family="Arial Bold"
    ),
    marker=dict(
        size=12,
        color=colors,
        line=dict(width=2, color='#2C3E50'),
        opacity=0.9,
        symbol='circle'
    ),
    name='Parole',
    hovertemplate='<b>%{text}</b><br>' +
                  'Vivente: %{x}<br>' +
                  'Dimensione: %{y}<br>' +
                  'Tecnologia: %{z}<br>' +
                  '<extra></extra>'
)

# Aggiunta griglia di riferimento sottile
def create_grid_lines():
    """Crea linee di griglia per facilitare la lettura"""
    grid_lines = []
    
    # Griglia sul piano XY (z=0)
    for i in np.arange(0, 1.3, 0.2):
        # Linee parallele all'asse X
        grid_lines.append(go.Scatter3d(
            x=[0, 1.2], y=[i, i], z=[0, 0],
            mode='lines', line=dict(color='#BDC3C7', width=1),
            showlegend=False, hoverinfo='skip', opacity=0.3
        ))
        # Linee parallele all'asse Y
        grid_lines.append(go.Scatter3d(
            x=[i, i], y=[0, 1.2], z=[0, 0],
            mode='lines', line=dict(color='#BDC3C7', width=1),
            showlegend=False, hoverinfo='skip', opacity=0.3
        ))
    
    return grid_lines

# Composizione della figura
fig = go.Figure(data=assi + axis_labels + [punti_principali] + create_grid_lines())

# Layout con impostazioni avanzate
fig.update_layout(
    title=dict(
        text='<b>Word Embeddings 3D</b><br><sub>Rappresentazione Semantica nello Spazio Vettoriale</sub>',
        x=0.5,
        font=dict(size=20, color='#2C3E50', family="Arial")
    ),
    width=900,   # Aumentata larghezza
    height=700,  # Aumentata altezza per bilanciare
    scene=dict(
        # Impostazioni assi migliorate
        xaxis=dict(
            title=dict(text='', font=dict(size=14, color='#E74C3C')),
            showspikes=False,
            showgrid=False,
            showline=False,
            zeroline=False,
            showticklabels=True,
            tickfont=dict(size=12, color='#34495E'),
            range=[0, 1.5]
        ),
        yaxis=dict(
            title=dict(text='', font=dict(size=14, color='#2ECC71')),
            showspikes=False,
            showgrid=False,
            showline=False,
            zeroline=False,
            showticklabels=True,
            tickfont=dict(size=12, color='#34495E'),
            range=[0, 1.5]
        ),
        zaxis=dict(
            title=dict(text='', font=dict(size=14, color='#3498DB')),
            showspikes=False,
            showgrid=False,
            showline=False,
            zeroline=False,
            showticklabels=True,
            tickfont=dict(size=12, color='#34495E'),
            range=[0, 1.5]
        ),
        # Impostazioni camera per vista ottimale
        camera=dict(
            eye=dict(x=1.8, y=1.8, z=1.5),
            center=dict(x=0, y=0, z=0),
            up=dict(x=0, y=0, z=1)
        ),
        # Sfondo e illuminazione
        bgcolor='#F8F9FA',
        aspectmode='cube'  # Mantiene proporzioni uguali
    ),
    # Margini ottimizzati
    margin=dict(l=20, r=20, b=20, t=80),
    # Sfondo generale
    paper_bgcolor='white',
    plot_bgcolor='white',
    # Font globale
    font=dict(family="Arial", size=12, color='#2C3E50')
)

# Aggiunta annotazioni esplicative
fig.add_annotation(
    x=0.02, y=0.98,
    xref="paper", yref="paper",
    text="<b>Legenda:</b><br>" +
         "🔴 Animali viventi<br>" +
         "🔵 Oggetti tecnologici<br>" +
         "⚪ Oggetti inanimati",
    showarrow=False,
    font=dict(size=11, color='#2C3E50'),
    bgcolor='rgba(255,255,255,0.8)',
    bordercolor='#BDC3C7',
    borderwidth=1,
    align="left"
)

# Impostazioni interattività
fig.update_layout(
    scene=dict(
        dragmode='turntable',  # Rotazione fluida
    ),
    hovermode='closest'
)

print("🎨 Visualizzazione Word Embeddings 3D migliorata")
print("📊 Caratteristiche:")
print("   • Font ottimizzati e leggibili")
print("   • Colori semantici per categorie")
print("   • Griglia di riferimento")
print("   • Hover informativi")
print("   • Layout bilanciato")
print("   • Interattività migliorata")

fig.show()

# Opzione per salvare ad alta risoluzione
# fig.write_html("word_embeddings_3d.html")
# fig.write_image("word_embeddings_3d.png", width=1200, height=900, scale=2)

🎨 Visualizzazione Word Embeddings 3D migliorata
📊 Caratteristiche:
   • Font ottimizzati e leggibili
   • Colori semantici per categorie
   • Griglia di riferimento
   • Hover informativi
   • Layout bilanciato
   • Interattività migliorata


**Un altro esempio di dimensioni semantiche**

In [12]:
import plotly.graph_objects as go

# Nuove parole e coordinate: (Astrattezza, Emotività, Temporalità)
parole = {
    "amore"     : (1, 1, 0.5),
    "macchina"  : (0, 0, 0),
    "libertà"   : (1, 1, 1),
    "scuola"    : (0.5, 0.3, 0.6),
    "sogno"     : (1, 0.8, 1),
    "pane"      : (0, 0, 0),
    "nostalgia" : (1, 0.9, -1),  # Valore temporale negativo per il passato
}

x = [v[0] for v in parole.values()]
y = [v[1] for v in parole.values()]
z = [v[2] for v in parole.values()]
labels = list(parole.keys())

# Assi evidenziati in nero
assi = [
    go.Scatter3d(x=[0, 1.2], y=[0, 0], z=[0, 0], mode='lines', line=dict(color='black', width=4), showlegend=False),
    go.Scatter3d(x=[0, 0], y=[0, 1.2], z=[0, 0], mode='lines', line=dict(color='black', width=4), showlegend=False),
    go.Scatter3d(x=[0, 0], y=[0, 0], z=[-1.2, 1.2], mode='lines', line=dict(color='black', width=4), showlegend=False)
]

# Punti e parole
punti = go.Scatter3d(
    x=x, y=y, z=z,
    mode='markers+text',
    text=labels,
    textposition='top center',
    marker=dict(size=6, color='lightcoral', line=dict(width=1, color='darkred'))
)

fig = go.Figure(data=assi + [punti])
fig.update_layout(
    title='Word Embedding 3D: Astrattezza, Emotività, Temporalità (con Passato)',
    scene=dict(
        xaxis_title='Astrattezza',
        yaxis_title='Emotività',
        zaxis_title='Temporalità',
        xaxis=dict(showspikes=False),
        yaxis=dict(showspikes=False),
        zaxis=dict(showspikes=False)
    ),
    margin=dict(l=0, r=0, b=0, t=50)
)

fig.show()


**Esempio di Operazioni Vettoriali sullo spazio delle Parole**

In [6]:
import plotly.graph_objects as go

# Vettori (Longitudine, Influenza, Cultura)
'''
punti = {
    "Italia": (0.6, 0.8, 0.9),
    "Roma": (0.6, 0.8, 0.88),
    "Francia": (0.5, 1.0, 1.0),
    "Parigi": (0.5, 1.0, 0.98),
    "USA": (0.2, 1.0, 1.0),
    "Washington": (0.2, 1.0, 0.98),
    "Russia": (0.9, 0.9, 0.6),
    "Mosca": (0.9, 0.9, 0.58),
    "Cina": (1.0, 1.0, 0.4),
    "Pechino": (1.0, 1.0, 0.38),
    "Giappone": (1.0, 0.9, 0.7),
    "Tokyo": (1.0, 0.9, 0.68),
    "Brasile": (0.3, 0.7, 0.8),
    "Brasilia": (0.3, 0.7, 0.78),
    "India": (0.95, 0.8, 0.5),
    "Nuova Delhi": (0.95, 0.8, 0.48)
}
'''

# longitudini normalizzate

punti = {
    "Italia": (0.6, 0.8, 0.9),
    "Roma": (0.413, 0.8, 0.88),
    "Francia": (0.5, 1.0, 1.0),
    "Parigi": (0.366, 1.0, 0.98),
    "USA": (0.2, 1.0, 1.0),
    "Washington": (0.0, 1.0, 0.98),
    "Russia": (0.9, 0.9, 0.6),
    "Mosca": (0.529, 0.9, 0.58),
    "Cina": (1.0, 1.0, 0.4),
    "Pechino": (0.893, 1.0, 0.38),
    "Giappone": (1.0, 0.9, 0.7),
    "Tokyo": (1.0, 0.9, 0.68),
    "Brasile": (0.3, 0.7, 0.8),
    "Brasilia": (0.134, 0.7, 0.78),
    "India": (0.95, 0.8, 0.5),
    "Nuova Delhi": (0.711, 0.8, 0.48)
}

x = [v[0] for v in punti.values()]
y = [v[1] for v in punti.values()]
z = [v[2] for v in punti.values()]
labels = list(punti.keys())

# Assi evidenziati
assi = [
    go.Scatter3d(x=[0, 1.1], y=[0, 0], z=[0, 0], mode='lines', line=dict(color='black', width=4), showlegend=False),
    go.Scatter3d(x=[0, 0], y=[0, 1.1], z=[0, 0], mode='lines', line=dict(color='black', width=4), showlegend=False),
    go.Scatter3d(x=[0, 0], y=[0, 0], z=[0, 1.1], mode='lines', line=dict(color='black', width=4), showlegend=False)
]

punti_fig = go.Scatter3d(
    x=x, y=y, z=z,
    mode='markers+text',
    text=labels,
    textposition='top center',
    marker=dict(size=6, color='lightblue', line=dict(width=1, color='darkblue'))
)

fig = go.Figure(data=assi + [punti_fig])
fig.update_layout(
    title='Embedding 3D di Nazioni e Capitali',
    scene=dict(
        xaxis_title='Localizzazione Geografica (Ovest → Est)',
        yaxis_title='Influenza Politica/Economica',
        zaxis_title='Stile Culturale (Tradizionale → Occidentale)',
        xaxis=dict(showspikes=False),
        yaxis=dict(showspikes=False),
        zaxis=dict(showspikes=False)
    ),
    margin=dict(l=0, r=0, b=0, t=50)
)

fig.show()


**Addizione e Sottrazione di Vettori**

$$Francia - Parigi = Italia - Roma$$

In [7]:
import numpy as np

In [8]:
Italia = np.array(punti['Italia'])
Italia

array([0.6, 0.8, 0.9])

In [9]:
Roma = np.array(punti['Roma'])
Roma

array([0.413, 0.8  , 0.88 ])

In [10]:
Francia = np.array(punti['Francia'])
Francia

array([0.5, 1. , 1. ])

In [11]:
Francia + (Roma - Italia)

array([0.313, 1.   , 0.98 ])

In [12]:
print(np.array(punti['Parigi']))

[0.366 1.    0.98 ]


**Prodotto di Vettori**

Il prodotto fra vettori può essere in qualche modo relazionato al concetto di similarità. Geometricamente il prodotto scalare misura quanto due vettori sono ***allineati***. Un prodotto scalare pari a zero indica che i vettori sono perpendicolari mentre un valore maggiore di zero indica un certo grado di allineamento nella stessa direzione e un numero minore di zero indica che sono più o meno allineati ma puntano in direzioni opposte. 

## Word2Vec

Un modello come **word2vec** impara gli embeddings (rappresentazioni numeriche delle parole) basandosi su un principio semplice:

> "**parole simili appaiono in contesti simili**".

Per esempio, le parole "gatto" e "cane" spesso compaiono vicine a parole come "animale", "cibo" o "giocare", mentre "auto" e "bicicletta" appaiono vicine a parole come "guidare", "strada", "velocità".

Ecco, passo per passo, come funziona in maniera molto semplificata:

1. **Scansione di un testo**:  
   Il modello legge tantissime frasi e osserva quali parole si trovano spesso vicine tra loro.

2. **Addestramento**:  
   Il modello cerca di prevedere, data una parola, quali parole tendono ad apparire vicine. Ad esempio, data la parola "gatto", il modello cerca di indovinare parole come "miagolare" o "croccantini".

3. **Conversione in numeri (embeddings)**:  
   Durante questo allenamento, il modello assegna automaticamente a ogni parola dei valori numerici (chiamati embedding) che catturano queste relazioni. Parole con significati o usi simili avranno embeddings numerici molto simili tra loro.

Alla fine, ogni parola è rappresentata da una sequenza di numeri (un vettore), che racchiude in sé il suo significato e i suoi rapporti con tutte le altre parole osservate.

Questo permette ai modelli di usare direttamente questi vettori numerici per calcolare similarità e analogie tra le parole, facendo operazioni matematiche molto semplici.

Il modello **word2vec** assegna a ciascuna parola dei valori numerici (embeddings) attraverso una procedura molto semplice, che si basa su tentativi ed errori e piccoli aggiustamenti continui:

Ecco, spiegato passo-passo con parole semplici:

### 1. **Inizia in modo casuale**
- All’inizio, ogni parola viene associata a una sequenza di numeri casuali.

### 2. **Gioco di "indovina la parola"**
- Il modello prende una parola dal testo, ad esempio *"gatto"*, e prova a prevedere quali altre parole spesso compaiono vicino ad essa (ad esempio "mangia", "gioca", "miagola").
- Se il modello riesce a prevedere correttamente le parole vicine, allora i numeri (embeddings) assegnati sono buoni. Se sbaglia, vuol dire che i numeri devono essere modificati.

### 3. **Impara dagli errori**
- Quando il modello sbaglia, cambia leggermente i valori numerici delle parole coinvolte per migliorare la previsione futura.
- Le parole che compaiono spesso insieme nel testo, progressivamente, avranno valori numerici sempre più simili tra loro.

### 4. **Ripeti molte volte**
- Il modello ripete questo processo moltissime volte, su milioni di frasi e parole.
- A ogni ripetizione, i valori numerici si aggiustano leggermente fino a diventare sempre più precisi.

### Alla fine del processo:
- Parole che appaiono in contesti simili finiscono per avere valori numerici molto vicini tra loro.
- Parole con significati molto diversi avranno valori numerici più distanti.

Questo meccanismo è ciò che permette a **word2vec** di catturare e rappresentare numericamente i significati e le relazioni tra le parole.

Certamente! Vediamo in modo più approfondito e intuitivo come funziona questa parte centrale del modello **word2vec**:

---

## 🟢 **Fase iniziale: parole e numeri casuali**

Immagina che ogni parola, all'inizio, sia associata a una sequenza casuale di numeri (un "embedding iniziale").

Ad esempio, semplificando:

- **gatto** → `[0.2, 0.6, 0.3]`
- **mangia** → `[0.8, 0.1, 0.5]`
- **cane** → `[0.4, 0.9, 0.7]`
- **miagola** → `[0.3, 0.2, 0.4]`

Questi numeri, inizialmente casuali, non significano ancora niente.

---

## 🔵 **Come funziona il gioco della previsione (training)**

Word2vec utilizza principalmente due strategie: vediamo la più comune, chiamata **skip-gram**.

Con la tecnica **skip-gram**, il modello segue questi passi:

### ① **Prende una parola centrale dal testo** (ad esempio "gatto").

Immagina la frase:

> "il **gatto** mangia e poi miagola"

La parola centrale scelta è **gatto**.

### ② **Prova a indovinare le parole vicine** (contesto).

In questa frase, intorno alla parola "gatto", ci sono parole vicine:  
- "il"  
- "mangia"  
- "e"  

Queste parole costituiscono il "contesto" di "gatto".

Il modello prende il vettore numerico (embedding) di **gatto** e cerca di predire (indovinare) gli embedding delle parole vicine.

### ③ **Calcola se ha indovinato o no (calcolo dell’errore)**.

In pratica, il modello converte gli embeddings in probabilità che ciascuna parola appaia nel contesto:

- se il modello dice correttamente che "mangia" o "miagola" sono vicine a "gatto", riceve un errore basso.
- se sbaglia (ad esempio dice che "automobile" o "televisore" sono vicine a "gatto"), l'errore è alto.

In questo passaggio, il modello utilizza una funzione matematica (solitamente una funzione chiamata "softmax" o versioni semplificate) per calcolare la probabilità di ogni parola di apparire vicino alla parola "gatto".

---

## 🟣 **Aggiornamento degli embeddings (apprendimento dagli errori)**

Una volta che il modello calcola quanto ha sbagliato, usa questo errore per modificare leggermente i numeri associati a "gatto" e alle parole vicine, per migliorare la prossima previsione.

- **Se ha predetto bene**: i numeri (embeddings) restano simili o subiscono modifiche piccolissime.
- **Se ha predetto male**: gli embeddings vengono modificati significativamente, per avvicinare "gatto" alle parole corrette e allontanarlo dalle parole sbagliate.

Ad esempio, se inizialmente "gatto" e "miagola" avevano numeri molto diversi tra loro, il modello, dopo aver notato che appaiono spesso insieme, modificherà i numeri per avvicinarli.

**Così, lentamente, embeddings casuali iniziali evolvono in embeddings significativi.**

---

## 🔴 **Iterazione del processo**

Questo processo viene ripetuto **milioni di volte**, passando attraverso un grande numero di parole e frasi diverse.

Ogni volta:

- il modello prende una nuova parola centrale.
- prova a indovinare le parole vicine.
- calcola l’errore.
- modifica gli embeddings in base a quell’errore.

Col passare del tempo, questo meccanismo produce embeddings numerici sempre più precisi e capaci di catturare il significato e la relazione tra parole.

---

## 🎯 **Cosa succede alla fine (risultato finale)?**

Alla fine dell’addestramento avrai una situazione tipo:

- **gatto** → `[0.2, 0.5, 0.4]`
- **miagola** → `[0.21, 0.51, 0.39]`
- **cane** → `[0.22, 0.52, 0.41]`

Noterai che "gatto", "miagola" e "cane" ora hanno numeri molto più simili tra loro, rispetto a parole diverse come "automobile" o "computer", che avranno numeri diversi.

Questi valori finali ("embeddings addestrati") rappresentano dunque il significato delle parole e i loro rapporti reciproci, **pronti per essere utilizzati per altre applicazioni** (ad esempio confrontare il significato di due parole o trovare sinonimi).

---

Questa è, in modo più dettagliato ma semplice, la procedura che permette al modello word2vec di assegnare a ogni parola valori numerici che catturano relazioni e significato.

## Uso degli embeddings

Come vengono utilizzati gli embeddings nel mondo dell'IA Generativa?

Partiamo con ChatGPT, che usa gli embeddings per capire il significato delle parole. Parola per parola, frase per frase per poi generare del testo. Noi abbiamo parlato di embeddings semplici a poche dimensioni, abbiamo parlato di Word2Vec che usa 300 dimensioni, ma giusto per darvi un'idea gli *embeddings* usati da ChatGPT usano **12888 dimensioni!**